[Model Element] Unload the model player (and its resources) when it's invisible

achan00 opened the pull request:
#### df1d0981bc3818eeb64d5369b4b30f7b834f976f
<pre>
[Model Element] Unload the model player (and its resources) when it&apos;s invisible
<a href="https://bugs.webkit.org/show_bug.cgi?id=292574">https://bugs.webkit.org/show_bug.cgi?id=292574</a>
<a href="https://rdar.apple.com/149972888">rdar://149972888</a>

Reviewed by Ryosuke Niwa and Mike Wyrzykowski.

When the model element&apos;s document has become hidden, unload its model player in
the model process to reduce its memory allocations. The unloading is done with
a delay in case the user quickly reactivates the page the model is in. The
unloading is handled by the model process, as the web content process will
likely get suspended when the page becomes hidden.

When the web content process is notified of the model player unloading, it
creates a PlaceholderModelPlayer with the current animation and transform
states. This way HTMLModelElement can continue to access/update that placeholder
model player. Any running animation timeline can keep running, and updates
can still be tracked by this placeholder model player.

When the page is shown again, HTMLModelElement gets the most current animation
and transform states from the placeholder model player, creates the &quot;real&quot;
model player, and restores it with those states. The reloading is also done
with a short delay in case the user quickly deactivates the page right after
activating it.

Unload the model player into a PlaceholderModelPlayer when the &lt;model&gt;
element is suspended when the page is added to the back/forward cache.
Any running animation is paused when the element is suspended, and
unpaused when the element is resumed.

* LayoutTests/model-element/model-element-on-hidden-page-expected.txt
* LayoutTests/model-element/model-element-on-hidden-page.html
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::visibilityStateChanged):
(WebCore::HTMLModelElement::unloadModelPlayer):
(WebCore::HTMLModelElement::reloadModelPlayer):
(WebCore::HTMLModelElement::reloadModelTimerFired):
(WebCore::HTMLModelElement::isVisible const):
(WebCore::HTMLModelElement::didUnload):
(WebCore::HTMLModelElement::renderingAbruptlyStopped): Deleted.
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::isPlaceholder const):
(WebCore::ModelPlayer::reload):
(WebCore::ModelPlayer::visibilityStateDidChange):
(WebCore::ModelPlayer::renderingAbruptlyStopped): Deleted.
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Modules/model-element/ModelPlayerTransformState.h:
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.cpp: Added.
(WebCore::PlaceholderModelPlayer::create):
(WebCore::PlaceholderModelPlayer::PlaceholderModelPlayer):
(WebCore::PlaceholderModelPlayer::currentAnimationState const):
(WebCore::PlaceholderModelPlayer::currentTransformState const):
(WebCore::PlaceholderModelPlayer::load):
(WebCore::PlaceholderModelPlayer::reload):
(WebCore::PlaceholderModelPlayer::boundingBoxCenter const):
(WebCore::PlaceholderModelPlayer::boundingBoxExtents const):
(WebCore::PlaceholderModelPlayer::entityTransform const):
(WebCore::PlaceholderModelPlayer::setEntityTransform):
(WebCore::PlaceholderModelPlayer::supportsTransform):
(WebCore::PlaceholderModelPlayer::setAutoplay):
(WebCore::PlaceholderModelPlayer::setLoop):
(WebCore::PlaceholderModelPlayer::setPlaybackRate):
(WebCore::PlaceholderModelPlayer::duration const):
(WebCore::PlaceholderModelPlayer::paused const):
(WebCore::PlaceholderModelPlayer::setPaused):
(WebCore::PlaceholderModelPlayer::currentTime const):
(WebCore::PlaceholderModelPlayer::setCurrentTime):
(WebCore::PlaceholderModelPlayer::setHasPortal):
(WebCore::PlaceholderModelPlayer::setStageMode):
(WebCore::PlaceholderModelPlayer::layer):
(WebCore::PlaceholderModelPlayer::layerHostingContextIdentifier):
(WebCore::PlaceholderModelPlayer::sizeDidChange):
(WebCore::PlaceholderModelPlayer::enterFullscreen):
(WebCore::PlaceholderModelPlayer::handleMouseDown):
(WebCore::PlaceholderModelPlayer::handleMouseMove):
(WebCore::PlaceholderModelPlayer::handleMouseUp):
(WebCore::PlaceholderModelPlayer::getCamera):
(WebCore::PlaceholderModelPlayer::setCamera):
(WebCore::PlaceholderModelPlayer::isPlayingAnimation):
(WebCore::PlaceholderModelPlayer::setAnimationIsPlaying):
(WebCore::PlaceholderModelPlayer::isLoopingAnimation):
(WebCore::PlaceholderModelPlayer::setIsLoopingAnimation):
(WebCore::PlaceholderModelPlayer::animationDuration):
(WebCore::PlaceholderModelPlayer::animationCurrentTime):
(WebCore::PlaceholderModelPlayer::setAnimationCurrentTime):
(WebCore::PlaceholderModelPlayer::hasAudio):
(WebCore::PlaceholderModelPlayer::isMuted):
(WebCore::PlaceholderModelPlayer::setIsMuted):
(WebCore::PlaceholderModelPlayer::accessibilityChildren):
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.cpp:
(WebKit::ModelConnectionToWebProcess::didUnloadModelPlayer):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.h:
* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.cpp:
(WebKit::ModelProcessModelPlayerManagerProxy::unloadModelPlayer):
* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::reloadModel):
(WebKit::ModelProcessModelPlayerProxy::modelVisibilityDidChange):
(WebKit::ModelProcessModelPlayerProxy::unloadModelTimerFired):
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::sizeDidChange):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Model/ModelProcessConnection.cpp:
(WebKit::ModelProcessConnection::didUnloadModelPlayer):
* Source/WebKit/WebProcess/Model/ModelProcessConnection.h:
* Source/WebKit/WebProcess/Model/ModelProcessConnection.messages.in:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::didUnload):
(WebKit::ModelProcessModelPlayer::reload):
(WebKit::ModelProcessModelPlayer::visibilityStateDidChange):
(WebKit::ModelProcessModelPlayer::sizeDidChange):
(WebKit::ModelProcessModelPlayer::renderingAbruptlyStopped): Deleted.
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.cpp:
(WebKit::ModelProcessModelPlayerManager::didUnloadModelProcessModelPlayer):
(WebKit::ModelProcessModelPlayerManager::modelProcessConnectionDidClose):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerTransformState.cpp:
(WebKit::ModelProcessModelPlayerTransformState::clone const):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerTransformState.h:

Canonical link: <a href="https://commits.webkit.org/294731@main">https://commits.webkit.org/294731@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3fdc72eb5caaae2797a213c1ae449b69bf53301

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102773 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22446 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12766 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107938 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53414 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104812 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22763 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30948 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78154 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35119 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105779 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17630 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92750 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58486 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17519 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10785 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52771 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87287 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10853 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110314 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29910 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22041 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87146 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30274 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88946 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86748 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22102 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31583 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9306 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24166 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29837 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35158 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29645 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32972 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31207 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -263,6 +268,7 @@ class HTMLModelElement final : public HTMLElement, private CachedRawResourceClie
     bool m_shouldCreateModelPlayerUponRendererAttachment { false };
 
     RefPtr<ModelPlayer> m_modelPlayer;
+    RunLoop::Timer m_reloadModelTimer;
```
We should be using WebCore's timer class here.
Even better, we can just post a task with delay using `EventLoop` object.

achan00 replied:
Fixed.

rniwa commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -339,6 +357,77 @@ void HTMLModelElement::deleteModelPlayer()
     m_modelPlayer = nullptr;
 }
 
+void HTMLModelElement::unloadModelPlayer()
```
We should probably run this logic right away when active DOM objects are getting suspended (i.e. page enters back/forward cache).

achan00 replied:
Fixed.

rniwa commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -108,6 +114,20 @@ Ref<HTMLModelElement> HTMLModelElement::create(const QualifiedName& tagName, Doc
+{
+    RELEASE_LOG(ModelElement, "%p - HTMLModelElement::suspend(): %d", this, static_cast<int>(reasonForSuspension));
+
+    if (reasonForSuspension == ReasonForSuspension::BackForwardCache)
```
We probably want to handle `ReasonForSuspension::PageWillBeSuspended` as well.

achan00 replied:
I saw that `HTMLMediaElement` does not pause playback with that reason - so I'll hold off on doing that here.

achan00 committed:
[Model Element] Unload the model player (and its resources) when it's invisible
https://bugs.webkit.org/show_bug.cgi?id=292574
rdar://149972888

Reviewed by Ryosuke Niwa and Mike Wyrzykowski.

When the model element's document has become hidden, unload its model player in
the model process to reduce its memory allocations. The unloading is done with
a delay in case the user quickly reactivates the page the model is in. The
unloading is handled by the model process, as the web content process will
likely get suspended when the page becomes hidden.

When the web content process is notified of the model player unloading, it
creates a PlaceholderModelPlayer with the current animation and transform
states. This way HTMLModelElement can continue to access/update that placeholder
model player. Any running animation timeline can keep running, and updates
can still be tracked by this placeholder model player.

When the page is shown again, HTMLModelElement gets the most current animation
and transform states from the placeholder model player, creates the "real"
model player, and restores it with those states. The reloading is also done
with a short delay in case the user quickly deactivates the page right after
activating it.

Unload the model player into a PlaceholderModelPlayer when the <model>
element is suspended when the page is added to the back/forward cache.
Any running animation is paused when the element is suspended, and
unpaused when the element is resumed.

* LayoutTests/model-element/model-element-on-hidden-page-expected.txt
* LayoutTests/model-element/model-element-on-hidden-page.html
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::visibilityStateChanged):
(WebCore::HTMLModelElement::unloadModelPlayer):
(WebCore::HTMLModelElement::reloadModelPlayer):
(WebCore::HTMLModelElement::reloadModelTimerFired):
(WebCore::HTMLModelElement::isVisible const):
(WebCore::HTMLModelElement::didUnload):
(WebCore::HTMLModelElement::renderingAbruptlyStopped): Deleted.
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::isPlaceholder const):
(WebCore::ModelPlayer::reload):
(WebCore::ModelPlayer::visibilityStateDidChange):
(WebCore::ModelPlayer::renderingAbruptlyStopped): Deleted.
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Modules/model-element/ModelPlayerTransformState.h:
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.cpp: Added.
(WebCore::PlaceholderModelPlayer::create):
(WebCore::PlaceholderModelPlayer::PlaceholderModelPlayer):
(WebCore::PlaceholderModelPlayer::currentAnimationState const):
(WebCore::PlaceholderModelPlayer::currentTransformState const):
(WebCore::PlaceholderModelPlayer::load):
(WebCore::PlaceholderModelPlayer::reload):
(WebCore::PlaceholderModelPlayer::boundingBoxCenter const):
(WebCore::PlaceholderModelPlayer::boundingBoxExtents const):
(WebCore::PlaceholderModelPlayer::entityTransform const):
(WebCore::PlaceholderModelPlayer::setEntityTransform):
(WebCore::PlaceholderModelPlayer::supportsTransform):
(WebCore::PlaceholderModelPlayer::setAutoplay):
(WebCore::PlaceholderModelPlayer::setLoop):
(WebCore::PlaceholderModelPlayer::setPlaybackRate):
(WebCore::PlaceholderModelPlayer::duration const):
(WebCore::PlaceholderModelPlayer::paused const):
(WebCore::PlaceholderModelPlayer::setPaused):
(WebCore::PlaceholderModelPlayer::currentTime const):
(WebCore::PlaceholderModelPlayer::setCurrentTime):
(WebCore::PlaceholderModelPlayer::setHasPortal):
(WebCore::PlaceholderModelPlayer::setStageMode):
(WebCore::PlaceholderModelPlayer::layer):
(WebCore::PlaceholderModelPlayer::layerHostingContextIdentifier):
(WebCore::PlaceholderModelPlayer::sizeDidChange):
(WebCore::PlaceholderModelPlayer::enterFullscreen):
(WebCore::PlaceholderModelPlayer::handleMouseDown):
(WebCore::PlaceholderModelPlayer::handleMouseMove):
(WebCore::PlaceholderModelPlayer::handleMouseUp):
(WebCore::PlaceholderModelPlayer::getCamera):
(WebCore::PlaceholderModelPlayer::setCamera):
(WebCore::PlaceholderModelPlayer::isPlayingAnimation):
(WebCore::PlaceholderModelPlayer::setAnimationIsPlaying):
(WebCore::PlaceholderModelPlayer::isLoopingAnimation):
(WebCore::PlaceholderModelPlayer::setIsLoopingAnimation):
(WebCore::PlaceholderModelPlayer::animationDuration):
(WebCore::PlaceholderModelPlayer::animationCurrentTime):
(WebCore::PlaceholderModelPlayer::setAnimationCurrentTime):
(WebCore::PlaceholderModelPlayer::hasAudio):
(WebCore::PlaceholderModelPlayer::isMuted):
(WebCore::PlaceholderModelPlayer::setIsMuted):
(WebCore::PlaceholderModelPlayer::accessibilityChildren):
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.cpp:
(WebKit::ModelConnectionToWebProcess::didUnloadModelPlayer):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.h:
* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.cpp:
(WebKit::ModelProcessModelPlayerManagerProxy::unloadModelPlayer):
* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::reloadModel):
(WebKit::ModelProcessModelPlayerProxy::modelVisibilityDidChange):
(WebKit::ModelProcessModelPlayerProxy::unloadModelTimerFired):
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::sizeDidChange):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Model/ModelProcessConnection.cpp:
(WebKit::ModelProcessConnection::didUnloadModelPlayer):
* Source/WebKit/WebProcess/Model/ModelProcessConnection.h:
* Source/WebKit/WebProcess/Model/ModelProcessConnection.messages.in:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::didUnload):
(WebKit::ModelProcessModelPlayer::reload):
(WebKit::ModelProcessModelPlayer::visibilityStateDidChange):
(WebKit::ModelProcessModelPlayer::sizeDidChange):
(WebKit::ModelProcessModelPlayer::renderingAbruptlyStopped): Deleted.
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.cpp:
(WebKit::ModelProcessModelPlayerManager::didUnloadModelProcessModelPlayer):
(WebKit::ModelProcessModelPlayerManager::modelProcessConnectionDidClose):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerTransformState.cpp:
(WebKit::ModelProcessModelPlayerTransformState::clone const):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerTransformState.h:

Canonical link: https://commits.webkit.org/294731@main
